### PR TITLE
Ios 623 - string functions

### PR DESCRIPTION
**Pull request by Poppy22**

### CODE REVIEW

File: src/Catty/Functions/Protocols/CBFunction.swift [hunk 1]
```diff
@@ -58,26 +58,46 @@ extension CBFunction {
...
+            return functionHeader       // no parameter function
         }
         
-        name += type(of: self).bracketOpen
+        functionHeader += type(of: self).bracketOpen
         for param in params {
-            name += param.defaultValueString()
+            
+            // add apostrophe at the beginning of a string
+            if ((functionName == "join" && (count == 0 || count == 1)) ||
```
Comment (m-herold) [inline src/Catty/Functions/Protocols/CBFunction.swift]: Please do not hardcode any function names here. Maybe you can add a new method to the _FunctionParameter_ directly and let the parameter decide whether to add apostrophes or not (can be added to every string parameter). Maybe something like _defaultValueDisplayString()_ ?

File: src/Catty/Functions/Text/LetterFunction.swift [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+    static func secondParameter() -> FunctionParameter {
+        return .string(defaultValue: "hello world")
+    }
+    
+    func value(firstParameter: AnyObject?, secondParameter: AnyObject?) -> String {
+        guard let number = firstParameter as? Int,
+            let text = secondParameter as? String else {
+                return type(of: self).defaultValue
+        }
+        let index = text.index(text.startIndex, offsetBy: number - 1)
```
Comment (m-herold) [inline src/Catty/Functions/Text/LetterFunction.swift]: According to Apple's documentation of the index function:

`/// The value passed as `n` must not offset `i` beyond the bounds of the collection.`

Please add a test where the first parameter is negative (e.g. -1). You will see that the app is crashing.

File: src/Catty/Functions/Text/JoinFunction.swift [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+    static func firstParameter() -> FunctionParameter {
+        return .string(defaultValue: "hello ")
+    }
+    
+    static func secondParameter() -> FunctionParameter {
+        return .string(defaultValue: "world")
+    }
+    
+    func value(firstParameter: AnyObject?, secondParameter: AnyObject?) -> String {
+        guard let firstText = firstParameter as? String,
```
Comment (m-herold) [inline src/Catty/Functions/Text/JoinFunction.swift]: If called this function with a number, e.g. join(1, 'hello') it should evaluate to the String '1hello', not '' (empty string). Please add a test for this scenario.

File: src/Catty/Functions/Text/JoinFunction.swift [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+    
+    static func firstParameter() -> FunctionParameter {
+        return .string(defaultValue: "hello ")
+    }
+    
+    static func secondParameter() -> FunctionParameter {
+        return .string(defaultValue: "world")
+    }
+    
+    func interpretParameter(parameter: AnyObject?) -> String? {
```
Comment (m-herold) [inline src/Catty/Functions/Text/JoinFunction.swift]: This function improves readability a lot and makes code shorter. Why are you returning an optional and force-unwrapping it later? Also, please make this function private.

File: src/Catty/Functions/Text/JoinFunction.swift [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+        if let text = parameter as? String {
+            return text
+        }
+        if let number = parameter as? Double {
+            if floor(number) == number {
+                return String(Int(number))
```
Comment (m-herold) [inline src/Catty/Functions/Text/JoinFunction.swift]: This Double to Int conversion can be very dangerous. Imaging the following input: `String(Int(Double.infinity))` which could result from `join('abc', 1/0)`. Please check this in all other Functions as well. Do not forget to write tests for these cases.
